### PR TITLE
support large schema discovery

### PR DESCRIPTION
**Pull request by mfsiega-airbyte**
## What
Support large catalog (aka schema) discovery.

## How
Today, the `discover schema` job passes the discovered schema back as a returned value. Then the API handler persists the schema and returns it (code ref: https://github.com/airbytehq/airbyte/blob/master/airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java#L210). For large schemas, this runs into a Temporal limit which won't pass such a large message.

Instead, we will persist the schema _in_ the `discover schema` job and just pass back the id. Then the handler will use the id to retrieve it out of the db and return it.

## Recommended reading order
1. `airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml`, modifies the output interface from the discover schema job.
2. `airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml`, `airbyte-config/config-models/src/main/resources/types/StandardDiscoverCatalogInput.yaml` modifies the input interface to the discover schema job, including some bits of info that are only necessary to persist alongside the catalog. This is probably not ideal, but could perhaps be cleaned up after the fix is merged.
3. `airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java` for changes in the API handler
4. `airbyte-workers/src/main/java/io/airbyte/workers/general/DefaultDiscoverCatalogWorker.java` where we actually persist the catalog
5. Everything else

## Tests

Updated unit tests to exercise the new behavior.

For manual e2e testing, used `make-big-schema.sh` to make a table with 15k tables. Verified that this failed under the existing behaviour and succeeded with this PR.

### CODE REVIEW

File: airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml
Comment (pmossman) [inline airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml:20]: Can we add a description to this, explaining how we determine the ID, if it's a UUID or not, etc?

File: airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml
Comment (pmossman) [inline airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml:25]: might be handy to add an example of what a "connectorVersion" string looks like here

File: airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml
Comment (pmossman) [inline airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml:28]: likewise, examples are always nice for generic 'string' type properties

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:258]: Now that this method is actually fetching from the database, I think a rename is appropriate so that it doesn't sound like it's doing a simple in-memory conversion operation

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:254]: It definitely feels like we've lost some readability with the change from `AirbyteCatalog` to `String` here. If the ID were a UUID, it'd be pretty clear, ie `SynchronousResponse<UUID>` but as a reader it feels like `String` could be anything. If this is actually a UUID, maybe we can find a way to type it as such so these response types are clear?

File: airbyte-server/src/main/java/io/airbyte/server/scheduler/DefaultSynchronousSchedulerClient.java
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/scheduler/DefaultSynchronousSchedulerClient.java:119]: This feels like the kind of thing that should maybe go into a helper or util class, and we can add some javadoc to explain why and how we're hashing the config. Otherwise I'd worry that this starts getting copy/pasted elsewhere and becomes hard to change if we ever need to

File: airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml
Comment (pmossman) [reply airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml:20]: Realized that maybe we can include a `format: uuid` here to get better autogenerated types throughout? Would be great to be able to say `SynchronousResponse<UUID>` instead of `SynchronousResponse<String>`!

File: airbyte-workers/src/test/java/io/airbyte/workers/general/DefaultDiscoverCatalogWorkerTest.java
Comment (pmossman) [inline airbyte-workers/src/test/java/io/airbyte/workers/general/DefaultDiscoverCatalogWorkerTest.java:157]: could probably delete these

**Review by pmossman — COMMENTED** [review]
@mfsiega-airbyte I gave this a first pass! I don't have a lot of hands-on experience with the Discover workflow, so I'm gonna tag in @davinchia to give this a review as well.

I think the approach looks good to me. Left a few suggestions around types, naming, etc.

For my own understanding, which parts of this were blocked on removing the catalog from the response of the connections list endpoint? I wanted to double check that blocker and make sure we're in the clear

File: airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml
Comment (davinchia) [inline airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml:19]: can we add a description here explaining what this id points to?

File: airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml
Comment (davinchia) [inline airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml:22]: in all the following, can we add to the description these fields are required in order to persist the result?

File: airbyte-config/config-models/src/main/resources/types/StandardDiscoverCatalogInput.yaml
Comment (davinchia) [inline airbyte-config/config-models/src/main/resources/types/StandardDiscoverCatalogInput.yaml:18]: just a comment we don't need to address in this PR: this is a smell that we need to wrap this twice.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:254]: I think it's a UUID, so we should be able to type it as such.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:258]: agree

File: airbyte-workers/src/main/java/io/airbyte/workers/general/DefaultDiscoverCatalogWorker.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/general/DefaultDiscoverCatalogWorker.java:93]: This should be moved to an api call in the future. Can we quickly write up an issue so we don't forget?

File: tools/bin/make-big-schema.sh
Comment (davinchia) [inline tools/bin/make-big-schema.sh:3]: nice!

File: tools/bin/make-big-schema.sh
Comment (davinchia) [inline tools/bin/make-big-schema.sh:10]: nit: newline

File: airbyte-server/src/main/java/io/airbyte/server/scheduler/DefaultSynchronousSchedulerClient.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/scheduler/DefaultSynchronousSchedulerClient.java:44]: note: there are now multiple instances of this floating about. We should centralise this. This can happen in a follow up PR. It definitely needs to happen otherwise we can bump heads against a tricky cache invalidation issues.

**Review by davinchia — COMMENTED** [review]
Thanks @mfsiega-airbyte , very exciting!

Most of my comments are in the same vein as Parker's. One major comment about centralising the hashing function definition. It doesn't have to happen in this PR but should definitely happen relatively soon after to prevent any cache invalidation errors.

I'm asking Malik to clarify here the hashing differences here: https://airbytehq-team.slack.com/archives/C03A8CSANPQ/p1664572102698309 

Will review again after you make changes.

File: airbyte-server/src/main/java/io/airbyte/server/scheduler/DefaultSynchronousSchedulerClient.java
Comment (mfsiega-airbyte) [reply airbyte-server/src/main/java/io/airbyte/server/scheduler/DefaultSynchronousSchedulerClient.java:119]: Yeah this is a good call - Davin opened https://github.com/airbytehq/airbyte/issues/17488.

File: airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml
Comment (mfsiega-airbyte) [reply airbyte-config/config-models/src/main/resources/types/ConnectorJobOutput.yaml:21]: Good call, done!

File: airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml
Comment (mfsiega-airbyte) [reply airbyte-config/config-models/src/main/resources/types/JobDiscoverCatalogConfig.yaml]: Wasn't exactly sure what an example of a hash should look like :) but I did put in some extra explanation?

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (mfsiega-airbyte) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java]: Attempted to make this a bit clearer (on top of the UUID change indeed).

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/discover/catalog/DiscoverCatalogActivityImpl.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/discover/catalog/DiscoverCatalogActivityImpl.java:81]: Do we need to version this in Temporal @pmossman ? Or is it ok to not do so because this is an activity impl and not a workflow?

**Review by davinchia — APPROVED** [review]
@mfsiega-airbyte this looks good to me.

One open question from me on versioning to make sure we don't cause errors. We can address the hashing in a follow up PR. 

Lmk if there are parts you want to me to look at closer.

**Comment by mfsiega-airbyte** [comment]
Re workflow versioning: once https://github.com/airbytehq/airbyte/pull/17562 is merged, any schema discovery workflows that are underway during a deployment will fail and need to be retried (rather than getting stuck).

Since this is relatively easy to recover from, planning to go ahead rather than specifically handling versioning (as soon as the linked PR is merged).

**Comment by mfsiega-airbyte** [comment]
Yeah indeed - these tests were never passing a `sourceId` which is a bit wrong since the field is required, but it wasn't actually being relied-upon before.

I'll send a PR to fix shortly.

**Comment by mfsiega-airbyte** [comment]
@suhomud PTAL: https://github.com/airbytehq/airbyte/pull/17662

**Pull request merged**
